Order Fulfillment: Pre-select the last provider used when adding a new shipment tracking

ctarda opened the pull request:
Fix #848 

This started as a simple change in the `ManualTrackingViewModel` and turned out being a new `Store`

First, the visuals:
![Screen Recording 2019-04-16 at 4 16 14 PM mov](https://user-images.githubusercontent.com/2722505/56193720-df8b7c00-6063-11e9-9857-9504c880d8be.gif)

## Changes:
### Storage:
- Added a protocol `FileStorage` and an implementation of it to read and write to a file on disk
- Added a model object `PreselectedProvider` to model the pair `siteID` and `providerName`

### Yosemite:
- Added a `AppSettingsStore` and `AppSettingsAction`. At the moment, the `Action` only exposed methods to save and retrieve a pair `ShipmentTrackingProvider` and `siteID`

### UI:
- Added a few lines to `AddTrackingViewModel` to user the `AppSettingsAction` to fetch and save the selected shipment tracking provider

## Testing
- Navigate to an order > Order fulfillment > Add Tracking
- Select a shipment tracking provider. 
- Add the tracking. or kill the app or dismiss the modal view.
- Tap Add Tracking again. The name of the provider should be populated with the last provider selected.
- Change stores, check that the name of the provider is different for different stores.

/ cc @jeanregisser 

Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

bummytime commented:
Oh...almost forgot! @ctarda be sure to merge `develop` into this PRs's branch and verify the build/tests are still happy in XCode 10.2. 😄 

Ref: #868

mindgraffiti commented on Yosemite/Yosemite/Stores/AppSettingsStore.swift:
```diff
@@ -0,0 +1,171 @@
+                                   onCompletion: onCompletion)
+        } catch {
+            let error = AppSettingsStoreErrors.parsePreselectedProvider
+            onCompletion(error)
```
Could we add a `DDLogError(error)` here, to give hints in the log that something went wrong?

mindgraffiti commented on Storage/StorageTests/Tools/FileStorageTests.swift:
```diff
@@ -0,0 +1,29 @@
+
+    override func tearDown() {
+        fileURL = nil
+        subject = nil
```
Does `super.tearDown()` belong here? I can't remember if it's required on tearDown().

ctarda replied:
🤦‍♂️  thanks!

mindgraffiti commented on Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift:
```diff
@@ -0,0 +1,143 @@
+        subject = nil
+        super.tearDown()
+    }
+
```
Can you also test that failures throw the expected error, such as `writePreselectedProvider`?

ctarda replied:
Thanks for catching that

ctarda commented on Yosemite/Yosemite/Stores/AppSettingsStore.swift:
```diff
@@ -0,0 +1,171 @@
+
+    /// URL to the plist file that we use to store the user selected
+    /// shipment tracing provider. Not declared as `private` so it can
+    /// be overriden in tests
```
Not nitpick at all, thanks!

ctarda commented:
Thanks for the review folks. I am pushing an update that addresses all your feedback.

The meatiest part of it is in commit 2326237d683996db7ec41cdbbbe709573089c1d3

I have added an Action to clear the stored providers, but in order to trigger it, what I have done is add an abstraction to represent the behaviour of being `Resettable`, and then make `AppSettingsStore` conform to it. When leaving the `AuthenticatedState`, we would execute that behaviour on all the `services` that implement it.

It seemed to be like it might be valuable, in the mid term, to do so. Let me know what you think!

bummytime commented:
Updates are working great @ctarda!

✅ Trackings can now be creating using the default provider
✅ Logging out clears the `plist` file
✅ Unit tests are all good

There is one change here that is good fodder for discussion — the `Resettable` protocol and conformance by `AppSettingsStore` (and more Stores eventually I assume). In this PR we have `AuthenticatedState` which is contained within the main Woo project calling the `reset()` func directly on Store(s) (via the protocol) in the Yosemite framework.

Thinking back to one of our architectural principals:
> Yosemite.Stores are *only expected process Actions*, and do not expose public API's to retrieve / observe objects. 

I know it's not explicitly stated there ☝️, but the original intent of `*Store`s was to only be indirectly "accessed" via actions from the main Woo app. Since we are bypassing the dispatcher and executing a func on the Store that mutates lower level data, I just wanted to understand the rationale a little better. Was there something specific where an action wouldn't work as well? (FWIW, I did notice you added `AppSettingsAction.resetStoredProviders` implementation). Just trying to make sure we are all on the same page 😄 

@mindgraffiti and @astralbodies any thoughts here?

mindgraffiti commented:
I'm not sure how to think in terms of architecture. I need to pause a minute and step back for a second.  Can you help me understand this @ctarda ? 

So if I was building this feature I would have used UserDefaults, then cleared it on log out or on store switch. I would use UserDefaults because saving the last used shipment provider seems like a user preference. What was the reasoning behind creating a new store?

ctarda commented:
@bummytime @mindgraffiti Let's begin with the rational for adding a store, as discussed in [this slack thread](p1555306667028700-slack-C6H8C3G23)

The different solutions considered were:
- Adding a couple of entities to the core data model.
- A plist file, containing an array of pairs (siteID, providerName), managed from  `AddManualTrackingViewModel`. In this context, "managed" means that the view model would load it, parse it, and find the relevant providerName per siteID, as well as save the selected provider to the plist file.
- Adding a new store. 

The reason to go with a store was that having a way to persist Application Settings (settings that are not synced through the API and are app specific, not store specific) seems to make sense. It is not only about persisting the selected provider, but also ranges in date selectors (which is already going to happen in a different feature). 

This way, the fact that a particular value is stored in a plist file, as well as the logic to read and write from it are completely abstracted out of the UI. The UI (in this particular case the view model backing it) does not need to know anything about how the data is persisted and does not need to know anything about how the pre-selected provider is picked.

When it comes to `Resettable`... yes, I am not so sure either. I added an Action to reset the tracking provider data, but when it came the moment to dispatch it, it kind of seemed to me like, given the fact that the `AuthenticatedState` already knows the services attached to it, it might make sense to introduce that extra abstraction ("being Resettable") because that also seems like a domain concern. Which, in a way, is compliant with the letter of the rules as _it does not not expose public API's to retrieve / observe objects_. 😛 

I will push a new commit to remove that abstraction and dispatch the action via the `StoresManager`. We can then compare and see what looks better?

ctarda commented:
@mindgraffiti @bummytime Done! `Resettable` is removed in the last commit and the action to reset and clear the saved providers is dispatched via the `StoresManager`. Please let me know what you think folks.

mindgraffiti commented:
Thank you for the long explanation @ctarda. I see where you're coming from now. 

You wrote an epic Slack thread and I missed it 😳 
Seriously though, P2 or it didn't happen 😉 

FWIW, I prefer the version without `Resettable`. Not because I have a sophisticated and well-reasoned answer, mostly because it "feels" somewhat off.

bummytime commented:
Merging this one for you @ctarda!
